build.gradle

```diff
@@ -59,57 +59,93 @@ task configureGroovyAll {
+            'MANIFEST_MF'                   : 'META-INF/MANIFEST.MF'
+        ]
+
+        def findDuplicates = { List<FileTree> fileForest ->
+            Map<String, List<File>> filesByRelativePath = [:].withDefault { [] }
+            fileForest*.visit([
+                visitDir : {},
+                visitFile: { filesByRelativePath[it.relativePath.toString()] << it.file }
+            ] as FileVisitor)
+            filesByRelativePath.findAll { it.value.size() > 1 }
+        }
 
-        tasks.groovyAll {
-            from writeVersionProperties
+        def concatenate = { List<File> files -> files*.text.join('\n') }
 
-            configurations.groovy.filter { it.name ==~ /groovy-.*\.jar/ }.each { groovyJar ->
-                def extractedFiles = zipTree(groovyJar)
+        def joinValues = { String properties, String keyName ->
+            (properties =~ /$keyName\s*=\s*(.+)/)*.getAt(1).join(',')
+        }
 
-                extractedFiles.filter { it.name == "org.codehaus.groovy.runtime.ExtensionModule" }.each { extensionModuleFile ->
-                    def extensionModule = new Properties()
-                    extensionModuleFile.withInputStream { input ->
-                        extensionModule.load(input)
-                    }
-                    extensionClasses << extensionModule.extensionClasses
-                    staticExtensionClasses << extensionModule.staticExtensionClasses
-                }
+        def mergeFiles = { String relative, List<File> files ->
+            switch (relative) {
+                case toMerge.LICENSE:
+                case toMerge.NOTICE:
+                case toMerge.AST_TRANSFORMATION:
+                    return concatenate(files)
+                case toMerge.GROOVY_RELEASE_INFO_PROPERTIES:
+                    return files[0].text
+                case toMerge.EXTENSION_MODULE:
+                    return """
+                        moduleName=groovy-all
+                        moduleVersion=${joinValues(files[0].text, 'moduleVersion')}
+                        extensionClasses=${joinValues(concatenate(files), 'extensionClasses')}
+                        staticExtensionClasses=${joinValues(concatenate(files), 'staticExtensionClasses')}
+                    """.stripIndent().trim()
+                case toMerge.MANIFEST_MF:
+                case toMerge.INDEX_LIST:
+                case toMerge.GROOVY_RUNNER:
+                    return null
+                default:
+                    throw new IllegalThreadStateException(files.toString())
+            }
+        }
 
-                from(extractedFiles) {
-                    exclude { it.path == "META-INF/groovy/org.codehaus.groovy.runtime.ExtensionModule" }
+        def mergeDuplicates = { FileTree tree, Map<String, List<File>> duplicates ->
+            def mergedFolder = new File(tree.tree.tmpDir, 'merged')
+            duplicates.each { relative, files ->
+                def mergedFile = new File(mergedFolder, relative)
```

Comment:
Hm, I remember adding a comment on this to say that it can be moved to the body of the `if` statement, where did that go? Not very important, but curious that it disappeared. :o

Reply:
Didn't see that comment, it's probably in another chrome tab (happened to me a few times :p)

Reply:
Done